### PR TITLE
CB-14568 CCM upgrade flow skeleton in Environment service

### PR DESCRIPTION
**Pull request by bergerdenes**
See detailed description in the commit message.

### CODE REVIEW

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java [hunk 1]
```diff
@@ -72,7 +72,18 @@
     LOAD_BALANCER_ENV_UPDATE_STARTED("Starting load balancer update for environment"),
     LOAD_BALANCER_ENV_UPDATE_FAILED("Failed to update environment with load balancer"),
     LOAD_BALANCER_STACK_UPDATE_STARTED("Starting load balancer update for all data lakes and data hubs"),
-    LOAD_BALANCER_STACK_UPDATE_FAILED("Failed to update data lakes and data hubs with load balancer");
+    LOAD_BALANCER_STACK_UPDATE_FAILED("Failed to update data lakes and data hubs with load balancer"),
+
+    UPGRADE_CCM_VALIDATION_IN_PROGRESS("Validation for Upgrade CCM started"),
```
Comment (lajosrodek) [inline environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java:77]: Remove "started".

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java [hunk 2]
```diff
@@ -72,7 +72,18 @@
...
+    UPGRADE_CCM_VALIDATION_FAILED("Validation for Upgrade CCM failed"),
+    UPGRADE_CCM_ON_FREEIPA_IN_PROGRESS("Upgrading CCM on Free IPA"),
+    UPGRADE_CCM_ON_FREEIPA_FAILED("Upgrading CCM on Free IPA failed"),
+    UPGRADE_CCM_ON_DATALAKE_IN_PROGRESS("Upgrading CCM on Data Lake"),
+    UPGRADE_CCM_ON_DATALAKE_FAILED("Upgrading CCM on Data Lake failed"),
+    UPGRADE_CCM_ON_DATAHUB_IN_PROGRESS("Upgrading CCM on Data Hub clusters"),
+    UPGRADE_CCM_ON_DATAHUB_FAILED("Upgrading CCM on Data Hub clusters failed"),
+    UPGRADE_CCM_FAILED("Upgrade CCM failed"),
+    UPGRADE_CCM_ROLLING_BACK("Rolling back failed CCM upgrade");
```
Comment (lajosrodek) [inline environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java:86]: Would it make sense to sync these messages with their counterparts in `messages.properties`?

File: environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java [hunk 3]
```diff
@@ -34,6 +34,11 @@
                 { "UPDATE_FAILED",                          EnvironmentStatus.UPDATE_FAILED,                        false },
                 { "ENVIRONMENT_RESOURCE_ENCRYPTION_INITIALIZATION_IN_PROGRESS", EnvironmentStatus.ENVIRONMENT_RESOURCE_ENCRYPTION_INITIALIZATION_IN_PROGRESS, false },
                 { "ENVIRONMENT_RESOURCE_ENCRYPTION_DELETE_IN_PROGRESS", EnvironmentStatus.ENVIRONMENT_RESOURCE_ENCRYPTION_DELETE_IN_PROGRESS, false },
+                { "UPGRADE_CCM_VALIDATION_FAILED",          EnvironmentStatus.UPGRADE_CCM_VALIDATION_FAILED,        false },
+                { "UPGRADE_CCM_ON_FREEIPA_FAILED",          EnvironmentStatus.UPGRADE_CCM_ON_FREEIPA_FAILED,        false },
+                { "UPGRADE_CCM_ON_DATALAKE_FAILED",         EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_FAILED,       false },
+                { "UPGRADE_CCM_ON_DATAHUB_FAILED",          EnvironmentStatus.UPGRADE_CCM_ON_DATAHUB_FAILED,        false },
+                { "UPGRADE_CCM_FAILED",                     EnvironmentStatus.UPGRADE_CCM_FAILED,                   false },
         };
```
Comment (lajosrodek) [inline environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java:42]: What about the new `*_IN_PROGRESS` and `UPGRADE_CCM_ROLLING_BACK` states?

File: environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java [hunk 4]
```diff
@@ -72,6 +77,13 @@ void isAvailableTest(String testCaseName, EnvironmentStatus underTest, boolean a
                 { "UPDATE_FAILED",                          EnvironmentStatus.UPDATE_FAILED,                        true },
                 { "ENVIRONMENT_RESOURCE_ENCRYPTION_INITIALIZATION_IN_PROGRESS", EnvironmentStatus.ENVIRONMENT_RESOURCE_ENCRYPTION_INITIALIZATION_IN_PROGRESS, false },
                 { "ENVIRONMENT_RESOURCE_ENCRYPTION_DELETE_IN_PROGRESS", EnvironmentStatus.ENVIRONMENT_RESOURCE_ENCRYPTION_DELETE_IN_PROGRESS, false },
+                { "FREEIPA_DELETED_ON_PROVIDER_SIDE",       EnvironmentStatus.FREEIPA_DELETED_ON_PROVIDER_SIDE, true },
+                { "UPGRADE_CCM_VALIDATION_FAILED",          EnvironmentStatus.UPGRADE_CCM_VALIDATION_FAILED, true },
+                { "UPGRADE_CCM_ON_FREEIPA_FAILED",          EnvironmentStatus.UPGRADE_CCM_ON_FREEIPA_FAILED, true },
+                { "UPGRADE_CCM_ON_DATALAKE_FAILED",         EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_FAILED, true },
+                { "UPGRADE_CCM_ON_DATAHUB_FAILED",          EnvironmentStatus.UPGRADE_CCM_ON_DATAHUB_FAILED, true },
+                { "UPGRADE_CCM_FAILED",                     EnvironmentStatus.UPGRADE_CCM_FAILED, true },
+
```
Comment (lajosrodek) [inline environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java:86]: What about the new `*_IN_PROGRESS` and `UPGRADE_CCM_ROLLING_BACK` states?

File: environment/src/main/java/com/sequenceiq/environment/environment/EnvironmentStatus.java [hunk 5]
```diff
@@ -124,7 +124,19 @@
...
+    UPGRADE_CCM_VALIDATION_FAILED(com.sequenceiq.environment.api.v1.environment.model.response.EnvironmentStatus.UPGRADE_CCM_VALIDATION_FAILED),
+    UPGRADE_CCM_ON_FREEIPA_IN_PROGRESS(com.sequenceiq.environment.api.v1.environment.model.response.EnvironmentStatus.UPGRADE_CCM_ON_FREEIPA_IN_PROGRESS),
+    UPGRADE_CCM_ON_FREEIPA_FAILED(com.sequenceiq.environment.api.v1.environment.model.response.EnvironmentStatus.UPGRADE_CCM_ON_FREEIPA_FAILED),
+    UPGRADE_CCM_ON_DATALAKE_IN_PROGRESS(com.sequenceiq.environment.api.v1.environment.model.response.EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_IN_PROGRESS),
+    UPGRADE_CCM_ON_DATALAKE_FAILED(com.sequenceiq.environment.api.v1.environment.model.response.EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_FAILED),
+    UPGRADE_CCM_ON_DATAHUB_IN_PROGRESS(com.sequenceiq.environment.api.v1.environment.model.response.EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_IN_PROGRESS),
+    UPGRADE_CCM_ON_DATAHUB_FAILED(com.sequenceiq.environment.api.v1.environment.model.response.EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_FAILED),
```
Comment (lajosrodek) [inline environment/src/main/java/com/sequenceiq/environment/environment/EnvironmentStatus.java:137]: Copy paste errors:
* Replace `EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_FAILED` with `EnvironmentStatus.UPGRADE_CCM_ON_DATAHUB_IN_PROGRESS`.
* Replace `EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_FAILED` with `EnvironmentStatus.UPGRADE_CCM_ON_DATAHUB_FAILED`.

**Review by lajosrodek — COMMENTED** [review]
Initial comments, some further may follow.

File: cloud-aws-native/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/resource/instance/alarm/AwsNativeCloudWatchService.java [hunk 6]
```diff
@@ -73,7 +73,7 @@ public AwsNativeCloudWatchService(CommonAwsClient commonAwsClient) {
     public void addCloudWatchAlarmsForSystemFailures(CloudResource resource, String regionName, AwsCredentialView credentialView) {
         try {
             PutMetricAlarmRequest metricAlarmRequest = createPutMetricAlarmRequest(resource.getInstanceId(), regionName);
-            LOGGER.debug("The following cloudwatch alarm – for instanceId: {} – has created and about to put it on AWS side: [{}]",
+            LOGGER.debug("The following cloudwatch alarm - for instanceId: {} - has created and about to put it on AWS side: [{}]",
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply cloud-aws-native/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/resource/instance/alarm/AwsNativeCloudWatchService.java:76]: yes

File: common/src/main/resources/messages/messages.properties [hunk 7]
```diff
@@ -420,6 +420,18 @@ environment.loadbalancer.update.finished=Load balancer update finished
...
 
+environment.upgrade.ccm.validation.started=Validation for upgrading CCM started
+environment.upgrade.ccm.validation.failed=Validation for upgrading CCM failed
+environment.upgrade.ccm.freeipa.started=Upgrading CCM on FreeIPA started
+environment.upgrade.ccm.freeipa.failed=Upgrading CCM on FreeIPA failed
+environment.upgrade.ccm.datalake.started=Upgrading CCM on datalake started
+environment.upgrade.ccm.datalake.failed=Upgrading CCM on datalake failed
+environment.upgrade.ccm.datahub.started=Upgrading CCM on datahub started
+environment.upgrade.ccm.datahub.failed=Upgrading CCM on datahub failed
+environment.upgrade.ccm.rollback=Rolling back CCM upgrade on last failed component
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply common/src/main/resources/messages/messages.properties:431]: under "component" I mean FreeIPA, DL or DH

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java (hunk 2 above)
Comment (bergerdenes) [reply environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java:86]: sure, on the other hand I do not see the purpose of this redundancy :)

Comment (lajosrodek) [reply environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java:86]: Agreed, I do not see it either. :smile:

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+                EnvironmentStatus.UPGRADE_CCM_VALIDATION_IN_PROGRESS, ResourceEvent.ENVIRONMENT_UPGRADE_CCM_VALIDATION_STARTED);
+    }
+
+    @Test
+    void validationNoEnvironment() {
+        testNoEnvironment(underTest::upgradeCcmValidationAction, UPGRADE_CCM_VALIDATION_HANDLER.selector());
+    }
+
+    @Test
+    void upgradeFreeipaHappyPath() {
```
Comment (lajosrodek) [inline environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:197]: Use camel case for "Freeipa": `FreeIpa`.

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+        testDeleteActionHappyPath(underTest::upgradeCcmInFreeIpaAction, UPGRADE_CCM_FREEIPA_HANDLER.selector(),
+                EnvironmentStatus.UPGRADE_CCM_ON_FREEIPA_IN_PROGRESS, ResourceEvent.ENVIRONMENT_UPGRADE_CCM_ON_FREEIPA_STARTED);
+    }
+
+    @Test
+    void upgradeFreeipaNoEnvironment() {
```
Comment (lajosrodek) [inline environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:203]: Use camel case for "Freeipa": `FreeIpa`.

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+        testNoEnvironment(underTest::upgradeCcmInFreeIpaAction, UPGRADE_CCM_FREEIPA_HANDLER.selector());
+    }
+
+    @Test
+    void upgradeDatalakeHappyPath() {
```
Comment (lajosrodek) [inline environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:208]: Use camel case for "Datalake": `DataLake`.

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+                EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_IN_PROGRESS, ResourceEvent.ENVIRONMENT_UPGRADE_CCM_ON_DATALAKE_STARTED);
+    }
+
+    @Test
+    void upgradeDatalakeNoEnvironment() {
+        testNoEnvironment(underTest::upgradeCcmInDataLakeAction, UPGRADE_CCM_DATALAKE_HANDLER.selector());
+    }
+
+    @Test
+    void upgradeDatahubHappyPath() {
```
Comment (lajosrodek) [inline environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:219]: Use camel case for "Datahub": `DataHub`.

**Review by lajosrodek — COMMENTED** [review]
Some further comments.

File: common/src/main/resources/messages/messages.properties [hunk 12]
```diff
@@ -420,6 +420,18 @@ environment.loadbalancer.update.finished=Load balancer update finished
...
 environment.loadbalancer.update.failed=Load balancer update failed
 
+environment.upgrade.ccm.validation.started=Validation for upgrading CCM started
+environment.upgrade.ccm.validation.failed=Validation for upgrading CCM failed
+environment.upgrade.ccm.freeipa.started=Upgrading CCM on FreeIPA started
+environment.upgrade.ccm.freeipa.failed=Upgrading CCM on FreeIPA failed
+environment.upgrade.ccm.datalake.started=Upgrading CCM on datalake started
+environment.upgrade.ccm.datalake.failed=Upgrading CCM on datalake failed
+environment.upgrade.ccm.datahub.started=Upgrading CCM on datahub started
+environment.upgrade.ccm.datahub.failed=Upgrading CCM on datahub failed
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply common/src/main/resources/messages/messages.properties:430]: ok

File: common/src/main/resources/messages/messages.properties [hunk 13]
```diff
@@ -420,6 +420,18 @@ environment.loadbalancer.update.finished=Load balancer update finished
 environment.loadbalancer.update.success=Load balancer update successful
 environment.loadbalancer.update.failed=Load balancer update failed
 
+environment.upgrade.ccm.validation.started=Validation for upgrading CCM started
+environment.upgrade.ccm.validation.failed=Validation for upgrading CCM failed
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply common/src/main/resources/messages/messages.properties:424]: ok

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java (hunk 1 above)
Comment (bergerdenes) [reply environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java:77]: ok

File: environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java (hunk 4 above)
Comment (bergerdenes) [reply environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java:86]: done

File: environment/src/main/java/com/sequenceiq/environment/environment/EnvironmentStatus.java (hunk 5 above)
Comment (bergerdenes) [reply environment/src/main/java/com/sequenceiq/environment/environment/EnvironmentStatus.java:137]: nice catch

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/AbstractUpgradeCcmAction.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractUpgradeCcmAction.class);
+
+    protected AbstractUpgradeCcmAction(Class<P> payloadClass) {
+        super(payloadClass);
+    }
+
+    @Override
+    protected CommonContext createFlowContext(FlowParameters flowParameters, StateContext<UpgradeCcmState, UpgradeCcmStateSelectors> stateContext,
+            P payload) {
+        return new CommonContext(flowParameters);
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/main/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/AbstractUpgradeCcmAction.java:29]: leaving it as is

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActions.java [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+        };
+    }
+
+    @Bean(name = "UPGRADE_CCM_FAILED_STATE")
+    public Action<?, ?> failedAction() {
+        return new AbstractUpgradeCcmAction<>(UpgradeCcmFailedEvent.class) {
+            @Override
+            protected void doExecute(CommonContext context, UpgradeCcmFailedEvent payload, Map<Object, Object> variables) {
+                LOGGER.warn(String.format("Failed to upgrade CCM in environment '%s'. Status: '%s'.",
+                        payload.getEnvironmentDto(), payload.getEnvironmentStatus()), payload.getException());
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/main/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActions.java:114]: why not

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/handler/UpgradeCcmOnDatahubHandler.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+import com.sequenceiq.environment.environment.flow.upgrade.ccm.event.UpgradeCcmEvent;
+import com.sequenceiq.environment.environment.flow.upgrade.ccm.event.UpgradeCcmFailedEvent;
+import com.sequenceiq.environment.environment.flow.upgrade.ccm.event.UpgradeCcmStateSelectors;
+import com.sequenceiq.flow.reactor.api.event.EventSender;
+import com.sequenceiq.flow.reactor.api.handler.EventSenderAwareHandler;
+
+import reactor.bus.Event;
+
+@Component
+public class UpgradeCcmOnDatahubHandler extends EventSenderAwareHandler<EnvironmentDto> {
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/main/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/handler/UpgradeCcmOnDatahubHandler.java:20]: keeping as-is as I see added value in ExceptionCatcherEventHandler

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java (hunk 9 above)
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:203]: ok

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java (hunk 8 above)
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:197]: ok

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java (hunk 10 above)
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:208]: c'mon ;)

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+    private void setActionPrivateFields(Action<?, ?> action) {
+        ReflectionTestUtils.setField(action, null, runningFlows, FlowRegister.class);
+        ReflectionTestUtils.setField(action, null, eventBus, EventBus.class);
+        ReflectionTestUtils.setField(action, null, reactorEventFactory, ErrorHandlerAwareReactorEventFactory.class);
+        ReflectionTestUtils.setField(action, null, tracer, Tracer.class);
+    }
+
+    private void verifyFailureEvent() {
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:286]: now it is

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+    @Test
+    void finishedHappyPath() {
+        testDeleteActionHappyPath(underTest::finishedAction, FINALIZE_UPGRADE_CCM_EVENT.selector(),
+                EnvironmentStatus.AVAILABLE, ResourceEvent.ENVIRONMENT_UPGRADE_CCM_FINISHED);
+        verify(metricService).incrementMetricCounter(eq(MetricType.ENV_UPGRADE_CCM_FINISHED), any(EnvironmentDto.class));
+    }
+
+    @Test
+    void finishedNoEnvironment() {
+        testNoEnvironment(underTest::finishedAction, UPGRADE_CCM_DATAHUB_HANDLER.selector());
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:238]: true

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+        Action<?, ?> action = configureAction(deleteAction);
+
+        action.execute(stateContext);
+
+        verify(environmentService, never()).save(any(Environment.class));
+        verify(environmentService, never()).getEnvironmentDto(any(Environment.class));
+        verify(environmentResponseConverter, never()).dtoToSimpleResponse(any(EnvironmentDto.class));
+        verify(notificationService, never()).send(any(ResourceEvent.class), any(SimpleEnvironmentResponse.class), anyString());
+        verify(eventBus).notify(selectorArgumentCaptor.capture(), eventArgumentCaptor.capture());
+    }
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:251]: added

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 20]
```diff
@@ -0,0 +1,303 @@
...
+                EnvironmentStatus.AVAILABLE, ResourceEvent.ENVIRONMENT_UPGRADE_CCM_FINISHED);
+        verify(metricService).incrementMetricCounter(eq(MetricType.ENV_UPGRADE_CCM_FINISHED), any(EnvironmentDto.class));
+    }
+
+    @Test
+    void finishedNoEnvironment() {
+        testNoEnvironment(underTest::finishedAction, UPGRADE_CCM_DATAHUB_HANDLER.selector());
+    }
+
+    private void testNoEnvironment(Supplier<Action<?, ?>> deleteAction, String selector) {
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:241]: ok

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 21]
```diff
@@ -0,0 +1,303 @@
...
+        action.execute(stateContext);
+
+        verify(environmentService, never()).save(any(Environment.class));
+        verify(environmentService, never()).getEnvironmentDto(any(Environment.class));
+        verify(environmentResponseConverter, never()).dtoToSimpleResponse(any(EnvironmentDto.class));
+        verify(notificationService, never()).send(any(ResourceEvent.class), any(SimpleEnvironmentResponse.class), anyString());
+        verify(eventBus).notify(selectorArgumentCaptor.capture(), eventArgumentCaptor.capture());
+    }
+
+    private void testDeleteActionHappyPath(Supplier<Action<?, ?>> deleteAction,
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:253]: done

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java [hunk 22]
```diff
@@ -0,0 +1,303 @@
...
+
+        EnvironmentDto environmentDto = mock(EnvironmentDto.class);
+        when(environmentStatusUpdateService.updateEnvironmentStatusAndNotify(any(), any(), any(), any(), any())).thenReturn(environmentDto);
+
+        action.execute(stateContext);
+
+        verify(eventBus).notify(selectorArgumentCaptor.capture(), eventArgumentCaptor.capture());
+        verify(reactorEventFactory).createEvent(headersArgumentCaptor.capture(), payloadArgumentCaptor.capture());
+        verify(environmentStatusUpdateService).updateEnvironmentStatusAndNotify(any(), any(), eq(environmentStatus), eq(eventStarted), any());
+    }
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/UpgradeCcmActionsTest.java:268]: added

File: environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java (hunk 3 above)
Comment (bergerdenes) [reply environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java:42]: added

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java [hunk 23]
```diff
@@ -72,7 +72,18 @@
     LOAD_BALANCER_ENV_UPDATE_STARTED("Starting load balancer update for environment"),
     LOAD_BALANCER_ENV_UPDATE_FAILED("Failed to update environment with load balancer"),
     LOAD_BALANCER_STACK_UPDATE_STARTED("Starting load balancer update for all data lakes and data hubs"),
-    LOAD_BALANCER_STACK_UPDATE_FAILED("Failed to update data lakes and data hubs with load balancer");
+    LOAD_BALANCER_STACK_UPDATE_FAILED("Failed to update data lakes and data hubs with load balancer"),
+
+    UPGRADE_CCM_VALIDATION_IN_PROGRESS("Validation for CCM upgrade"),
+    UPGRADE_CCM_VALIDATION_FAILED("Validation for CCM upgrade failed"),
+    UPGRADE_CCM_ON_FREEIPA_IN_PROGRESS("Upgrading CCM on Free IPA"),
+    UPGRADE_CCM_ON_FREEIPA_FAILED("Upgrading CCM on Free IPA failed"),
```
Comment (lajosrodek) [inline environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java:80]: Nit: remove space in "Free IPA".

File: environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java [hunk 24]
```diff
@@ -72,6 +82,17 @@ void isAvailableTest(String testCaseName, EnvironmentStatus underTest, boolean a
                 { "UPDATE_FAILED",                          EnvironmentStatus.UPDATE_FAILED,                        true },
                 { "ENVIRONMENT_RESOURCE_ENCRYPTION_INITIALIZATION_IN_PROGRESS", EnvironmentStatus.ENVIRONMENT_RESOURCE_ENCRYPTION_INITIALIZATION_IN_PROGRESS, false },
                 { "ENVIRONMENT_RESOURCE_ENCRYPTION_DELETE_IN_PROGRESS", EnvironmentStatus.ENVIRONMENT_RESOURCE_ENCRYPTION_DELETE_IN_PROGRESS, false },
+                { "FREEIPA_DELETED_ON_PROVIDER_SIDE",       EnvironmentStatus.FREEIPA_DELETED_ON_PROVIDER_SIDE,     true },
+                { "UPGRADE_CCM_VALIDATION_IN_PROGRESS",     EnvironmentStatus.UPGRADE_CCM_VALIDATION_IN_PROGRESS,   false },
+                { "UPGRADE_CCM_ON_FREEIPA_IN_PROGRESS",     EnvironmentStatus.UPGRADE_CCM_ON_FREEIPA_IN_PROGRESS,   false },
+                { "UPGRADE_CCM_ON_DATALAKE_IN_PROGRESS",    EnvironmentStatus.UPGRADE_CCM_ON_DATALAKE_IN_PROGRESS,  false },
+                { "UPGRADE_CCM_ON_DATAHUB_IN_PROGRESS",     EnvironmentStatus.UPGRADE_CCM_ON_DATAHUB_IN_PROGRESS,   false },
+                { "UPGRADE_CCM_ROLLING_BACK",               EnvironmentStatus.UPGRADE_CCM_ROLLING_BACK,             false },
```
Comment (lajosrodek) [inline environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java:90]: I wonder if the `UPGRADE_CCM_ROLLING_BACK` state should be considered a failed one. But perhaps not, I suppose the transition will look like e.g. `UPGRADE_CCM_ON_FREEIPA_IN_PROGRESS` -> `UPGRADE_CCM_ROLLING_BACK` -> `UPGRADE_CCM_ON_FREEIPA_FAILED`, right?

**Review by lajosrodek — APPROVED** [review]
Thanks for the fixes. Two minor comments, but looks great otherwise. :smile:

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java (hunk 23 above)
Comment (bergerdenes) [reply environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatus.java:80]: there are multiple occurrences with this wording, I correct them too

File: environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java (hunk 24 above)
Comment (bergerdenes) [reply environment-api/src/test/java/com/sequenceiq/environment/api/v1/environment/model/response/EnvironmentStatusTest.java:90]: this is not yet planned but yes it seems a good idea.

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/handler/UpgradeCcmOnDatalakeHandler.java [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+    public String selector() {
+        return UPGRADE_CCM_DATALAKE_HANDLER.name();
+    }
+
+    @Override
+    public void accept(Event<EnvironmentDto> environmentDtoEvent) {
+        LOGGER.debug("In UpgradeCcmOnDatalakeHandler.accept");
+        EnvironmentDto environmentDto = environmentDtoEvent.getData();
+        try {
+            // TODO action here
```
Comment (lnardai) [inline environment/src/main/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/handler/UpgradeCcmOnDatalakeHandler.java:38]: Is this still relevant?

**Review by lnardai — APPROVED** [review]

**Pull request merged**